src/app/clusters/concentration-measurement-server/concentration-measurement-server.cpp

```diff
@@ -0,0 +1,512 @@
+        }
+
+        if (!mMinMeasuredValue.IsNull())
+        {
+            if (aAverageMeasuredValue.Value() < mMinMeasuredValue.Value())
+            {
+                return CHIP_ERROR_INVALID_ARGUMENT;
+            }
+        }
+    }
+
+    DataModel::Nullable<float> oldValue = mAverageMeasuredValue;
+    mAverageMeasuredValue               = aAverageMeasuredValue;
+
+    if (oldValue != mAverageMeasuredValue)
+    {
+        ConcreteAttributePath path = ConcreteAttributePath(mEndpointId, mClusterId, Attributes::AverageMeasuredValue::Id);
+        MatterReportingAttributeChangeCallback(path);
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Instance::SetAverageMeasuredValueWindow(uint32_t aAverageMeasuredValueWindow)
+{
+    if (!HasFeature(Feature::kAverageMeasurement))
+    {
+        return CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE;
+    }
+
+    if (aAverageMeasuredValueWindow > WINDOW_MAX)
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    uint32_t oldValue           = mAverageMeasuredValueWindow;
+    mAverageMeasuredValueWindow = aAverageMeasuredValueWindow;
+
+    if (oldValue != mAverageMeasuredValueWindow)
+    {
+        ConcreteAttributePath path = ConcreteAttributePath(mEndpointId, mClusterId, Attributes::AverageMeasuredValueWindow::Id);
+        MatterReportingAttributeChangeCallback(path);
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Instance::SetUncertainty(float aUncertainty)
+{
+    if (!HasFeature(Feature::kNumericMeasurement))
+    {
+        return CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE;
+    }
+
+    float oldValue = mUncertainty;
+    mUncertainty   = aUncertainty;
+
+    if (oldValue != mUncertainty)
+    {
+        ConcreteAttributePath path = ConcreteAttributePath(mEndpointId, mClusterId, Attributes::Uncertainty::Id);
+        MatterReportingAttributeChangeCallback(path);
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Instance::SetLevel(LevelValueEnum aLevel)
+{
+    if (!HasFeature(Feature::kLevelIndication))
+    {
+        return CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE;
+    }
+
+    if (aLevel == LevelValueEnum::kUnknownEnumValue)
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    if (!HasFeature(Feature::kMediumLevel) && aLevel == LevelValueEnum::kMedium)
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    if (!HasFeature(Feature::kCriticalLevel) && aLevel == LevelValueEnum::kCritical)
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    LevelValueEnum oldValue = mLevel;
+    mLevel                  = aLevel;
+
+    if (oldValue != mLevel)
+    {
+        ConcreteAttributePath path = ConcreteAttributePath(mEndpointId, mClusterId, Attributes::LevelValue::Id);
+        MatterReportingAttributeChangeCallback(path);
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Instance::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    switch (aPath.mAttributeId)
+    {
+    case Attributes::MeasuredValue::Id:
```

Comment:
So here's an issue with doing things this way.  If I am a cluster with the LEV feature.... I am now paying the storage an codesize cost for all of this other stuff that's only allowed for the MEA feature, because it will _all_ get pulled in here.

This is really rather poor.

Note that just creating an API does not require switching to this setup and its attendant bloat.... And yes, we have multiple concentration measurement things here in all-clusters-app, but did you notice that we are picking up 4KB of flash here on the embedded platforms?

Reply:
> but did you notice that we are picking up 4KB of flash here on the embedded platforms?

@bzbarsky-apple :disappointed: I didn't notice that!

Would it be acceptable to just abstract `emberAfWriteAttribute`/`emberAfReadAttribute` here instead?

Reply:
Or maybe we just do nothing - there's zero extra cost to using the accessors as they are, given we have no indication we'll move away from these accessors anytime fast?

Reply:
That would be just fine, yes.  Basically, have the API you have now but just store in the attribute store.  And then if we decide to change the implementation later we can.